fdmi/plugins/fdmi_app

```diff
@@ -60,19 +62,119 @@ MAX_KEYSTRS = 10240
 # is communicating.
 processes = []
 
+FDMI_producer = None
+ack_thread = None
+lock = threading.Lock()
+
+
+def send_ack_to_plugin(banner, p, rec_id):
+    print(banner + " ACK " + rec_id, flush = True)
+    ri = bytes('{}\n'.format(rec_id), encoding = "utf-8")
+    lock.acquire()
+    p.stdin.write(ri)
+    p.stdin.flush()
+    lock.release()
+
+
+
+class Acker (threading.Thread):
+    def __init__(self, threadID, name, args):
+        threading.Thread.__init__(self)
+        self.threadID = threadID
+        self.name = name
+        self.args = args
+        self.consumer = None
+        self.done = False
+
+    def run(self):
+        if self.args.kafka_server is not None:
+            print("Starting: " + self.name, flush = True)
+            self.consume_and_ack()
+
+    def terminate(self):
+        if self.args.kafka_server is not None:
+            self.done = True
+            print("Flagging " + self.name + " Done", flush = True)
+
+    def consume_and_ack(self):
+        from kafka import KafkaConsumer
+
+        self.consumer = KafkaConsumer("FDMI",
+                                 bootstrap_servers=self.args.kafka_server,
+                                 group_id='acker'+self.name,
+                                 auto_offset_reset='earliest',
+                                 api_version=(2,13))
+        while True :
+            msgs = self.consumer.poll(500)
+            if msgs is not None and len(msgs) != 0:
+                for records in msgs.values():
+                    for rec in records:
+                        kvs = json.loads(rec.value.decode())
+                        for p in processes:
+                            rec_id = kvs.get('rec_id')
+                            send_ack_to_plugin("----------", p, rec_id)
+
+            if self.done and len(msgs) == 0:
+                break
+
+        self.consumer.close()
+        print("Acker thread loop done", flush = True)
+
+
+
+
+# Establish connection to kafka server
+def kafka_producer_connect(args):
+    global FDMI_producer
+
+    if (args.kafka_server is not None):
+        from kafka import KafkaProducer
+        FDMI_producer = KafkaProducer(bootstrap_servers=args.kafka_server,
+                         value_serializer=lambda m: json.dumps(m).encode())
+
+        if (not FDMI_producer.bootstrap_connected()):
+            print('Can not connect to kafka server: {}'.format(args.kafka_server),
+                  file=sys.stderr)
+            sys.exit(einval)
+
+
+# Close connection to kafka server
+def kafka_producer_close():
+    global FDMI_producer
+
+    if (FDMI_producer is not None):
+        FDMI_producer.flush()
+        FDMI_producer.close()
+        FDMI_producer = None
+
+
+# Send FDMI record to kafka with "FDMI" topic
+def kafka_producer_send(kv):
+    global FDMI_producer
+
+    if (FDMI_producer is not None):
+        future = FDMI_producer.send("FDMI", kv)
```

Comment:
where is FDMI_producer.send()?

Reply:
Send message to kafka server.
And Replication Manager can receive these messages from the same kafka server, and process them.

Reply:
is this FDMI_producer an external library?

Reply:
No. It is just an variable, returned from connection.